Evaluate all dependencies for an editor to determine visibility

mstokes-f5 opened the pull request:
Fixes #703 by not evaluating dependencies independently.

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks backward-compatibility?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | #703
| Updated README/docs?   | ❌
| Added CHANGELOG entry?   | ✔️

It is possible that users were relying on the previous behavior. This would be unreliable since the visibility is determined by the last dependency evaluated, which would be determined by the order of the watch events. I am not sure this order is documented/guaranteed.

Also, `checkDependency()` was left in place to minimize API changes, but it no longer behaves the same. It no longer calls `notify()` or updates visibility. Also, the state of `dependenciesFulfilled` will  likely be different when calling this function with and without this PR.

schmunk42 commented:
I have migrated the repository to travis-ci.com to fix the missing CI status, could you recheck this PR, since there are conflicts popping up now. 

I am not sure why they are showing up. Sorry for the inconvenience.

germanbisurgi commented:
This is a complex issue. I played around using @mstokes-f5 brach and the current master branch with his test schema. I notice that the bug was fixed. Also run the tests in local and every test was green. I am not sure if this PR breaks backward-compatibility. It seems it does not but maybe there are not sufficient tests written for this feature.

In conclusion, i think we should write a test for it and just for security, made this change available in the next release (in case it PR breaks backward-compatibility)

mstokes-f5 commented:
Sorry for the long radio silence on this; some things kept me away from this. However, I am able to work on this again now.

I agree that there likely isn't sufficient test coverage around this to check for potential backward-compatibility-breaking issues. Did you have something in mind for a test, or should I just play around a bit and propose something?

Are you concerned about the behavior change around `checkDependency()` and `dependenciesFulfilled` mentioned in the original post?

mstokes-f5 commented:
I have rebased this on to the current `master` (2800a45eadd6c1727bd1170f6c09c47f0a3133e9) to resolve the merge conflict. I left the `memo` commit in place instead of squashing since I was uncertain if the commit message is part of any process.

schmunk42 commented:
ping @germanbisurgi

germanbisurgi commented:
I still think tests are needed

mstokes-f5 commented:
I have rebased this on the current `master` branch and added a new end-to-end Scenario. The new test is based on the sample provided in #703 and demonstrates that this PR fixes the issue. Please let me know what further tests you would like to see for this PR.

schmunk42 commented:
Thank you!
